surface.c

```diff
@@ -0,0 +1,38 @@
+#define _POSIX_C_SOURCE 200809L
+#include <string.h>
+#include <stdlib.h>
+
+#include "config.h"
+#include "mako.h"
+#include "surface.h"
+
+void destroy_surface(struct mako_surface *surface) {
+	if (surface->layer_surface != NULL) {
+		zwlr_layer_surface_v1_destroy(surface->layer_surface);
+	}
+	if (surface->surface != NULL) {
+		wl_surface_destroy(surface->surface);
+	}
+	finish_buffer(&surface->buffers[0]);
+	finish_buffer(&surface->buffers[1]);
+
+	/* Clean up memory resources */
+	free(surface->name);
+	wl_list_remove(&surface->link);
+	free(surface);
+}
+
+struct mako_surface *create_surface(struct mako_state *state,
+		struct mako_surface_config *config) {
+	struct mako_surface *surface = calloc(1, sizeof(*surface));
+	if (!surface) {
+		return NULL;
+	}
+
+	surface->config = config;
+	surface->name = strdup(config->name);
```

Comment:
I don't think we should copy this since the config should always outlive any surface using it.

Reply:
I thought so as well, until I got to reapply_config part.
That invalidates all configs, but not the surfaces currently.

I could tear down all surfaces and re-create them again, but I felt that it made more sense to keep them, and just cleanup. Both from ease of adding it to codeflow, and a (micro)optimization.

Reply:
Hmm, I see. We do our best to immediately apply changes to styles, I wonder if this should include moving notifications? If so then we'd need to tear them all down in case things moved.

Reply:
Hm, I think we are thinking about slightly different things.

So the current code order is:
1. Deallocate configs
2. Re-read configuration
3. Ensure every configured surface exists
4. Reapply styles, which re-evaluates the surface
5. Trigger redraw / de-allocate unused surfaces

The new step is 3, and the added parts to 4 and 5 of course.
Everything still happens at the same time. They don't change the timing (outside of course pure computation time) at all.